import into failed and tidb panic when inject pd leader network partition

## Bug Report

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What did you do?
1、run import into
2、inject pd leader network partition

### What did you expect to see?
import into can success

### What did you see instead?
import into failed

the status of import job is not finished or running (now: 2024-01-05 06:36:06, jobId: 90002, step: importing, status: failed)
operatorLogs:
[2024-01-05 06:10:35] ###### start import into
[2024-01-05 06:10:35] ###### wait for import job to finish
[2024-01-05 06:36:06] ###### wait for import job to finish failed
select id, step, status from mysql.tidb_import_jobs where start_time >= '2024-01-05 06:10:35'
jobId: 90002, step: importing, status: failed"

tidb logs
[tidb-0-500.txt](https://github.com/tikv/pd/files/13856100/tidb-0-500.txt)
[tidb-1-500.txt](https://github.com/tikv/pd/files/13856070/tidb-1-500.txt)

tidb-1：
[2024/01/05 06:35:55.885 +08:00] [ERROR] [handle.go:92] ["task reverted"] [task-id=90002] [error="[0]runtime error: invalid memory address or nil pointer dereference"]
[2024/01/05 06:35:56.104 +08:00] [INFO] [scheduler_manager.go:286] ["task finished"] [task-id=90002]
[2024/01/05 06:35:56.107 +08:00] [INFO] [scheduler_manager.go:329] ["cleanUp routine start"]
[2024/01/05 06:35:56.115 +08:00] [INFO] [scheduler_manager.go:338] ["cleanUp routine success"]
[2024/01/05 06:36:09.225 +08:00] [WARN] [info.go:473] ["response not 200"] [method=GET] [host=[tc-pd-1.tc](http://tc-pd-1.tc/)-pd-peer.endless-ha-test-import-into-tps-5490660-1-730.svc:2379] [url=/pd/api/v1/stores] ["http status"=500]
[2024/01/05 06:36:09.226 +08:00] [ERROR] [client.go:245] ["[pd] request failed with a non-200 status"] [source=tikv-driver] [name=GetMinResolvedTSByStoresIDs] [url=http://tc-pd-1.tc-pd-peer.endless-ha-test-import-into-tps-5490660-1-730.svc:2379/pd/api/v1/min-resolved-ts] [method=GET] [caller-id=pd-http-client] [status="500 Internal Server Error"] [body="[PD:apiutil:ErrRedirectToNotLeader]redirect to not leader\n"]

tidb-0：
[2024/01/05 06:35:47.111 +08:00] [ERROR] [wait_group_wrapper.go:232] ["panic in error group"] [recover="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).generateAndSendJob.(*ErrorGroupWithRecover).Go.func3.1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/wait_group_wrapper.go:232\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:914\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:261\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:861\ngithub.com/tikv/pd/client.(*client).ScanRegions\n\t/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20240103101103-a4d2f1ca365a/client.go:1040\ngithub.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).ScanRegions\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/restore/split/client.go:514\ngithub.com/pingcap/tidb/br/pkg/restore/split.PaginateScanRegion.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/restore/split/split.go:106\ngithub.com/pingcap/tidb/br/pkg/utils.WithRetry.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/utils/retry.go:213\ngithub.com/pingcap/tidb/br/pkg/utils.WithRetryV2[...]\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/utils/retry.go:231\ngithub.com/pingcap/tidb/br/pkg/utils.WithRetry\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/utils/retry.go:212\ngithub.com/pingcap/tidb/br/pkg/restore/split.PaginateScanRegion\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/restore/split/split.go:101\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).generateJobForRange\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:978\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).generateAndSendJob.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/lightning/backend/local/local.go:899\ngithub.com/pingcap/tidb/br/pkg/lightning/backend/local.(*Backend).generateAndSendJob.(*ErrorGroupWithRecover).Go.func3\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/pkg/util/wait_group_wrapper.go:236\ngolang.org/x/sync/errgroup.(*Group).Go.func1\n\t/go/pkg/mod/golang.org/x/sync@v0.5.0/errgroup/errgroup.go:75"]
[2024/01/05 06:35:54.923 +08:00] [ERROR] [manager.go:438] ["failed to handle task"] [error="runtime error: invalid memory address or nil pointer dereference"]

runtime.gopanic
    /usr/local/go/src/runtime/panic.go:914
runtime.panicmem
    /usr/local/go/src/runtime/panic.go:261
runtime.sigpanic
    /usr/local/go/src/runtime/signal_unix.go:861
github.com/tikv/pd/client.(*client).ScanRegions
    /go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20240103101103-a4d2f1ca365a/client.go:1040
github.com/pingcap/tidb/br/pkg/restore/split.(*pdClient).ScanRegions
    /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/br/pkg/restore/split/client.go:514

### What version of PD are you using (`pd-server -V`)?
./pd-server -V
 Release Version: v7.6.0-alpha
Edition: Community
Git Commit Hash: 754095a9f460dcf31f053045cfedfb00b9ad8e81
Git Branch: heads/refs/tags/v7.6.0-alpha
UTC Build Time:  2024-01-04 11:37:28
2024-01-05T06:36:50.031+0800

--- a/client/client.go
+++ b/client/client.go
@@ -25,6 +25,7 @@ import (
 
 	"github.com/opentracing/opentracing-go"
 	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
 	"github.com/pingcap/kvproto/pkg/metapb"
 	"github.com/pingcap/kvproto/pkg/pdpb"
 	"github.com/pingcap/log"
@@ -1042,6 +1043,9 @@ func (c *client) ScanRegions(ctx context.Context, key, endKey []byte, limit int,
 		return nil, errs.ErrClientGetProtoClient
 	}
 	resp, err := pdpb.NewPDClient(serviceClient.GetClientConn()).ScanRegions(cctx, req)
+	failpoint.Inject("responseNil", func() {
+		resp = nil
+	})
 	if serviceClient.NeedRetry(resp.GetHeader().GetError(), err) {
 		protoClient, cctx := c.getClientAndContext(scanCtx)
 		if protoClient == nil {

--- a/client/client.go
+++ b/client/client.go
@@ -1042,7 +1042,7 @@ func (c *client) ScanRegions(ctx context.Context, key, endKey []byte, limit int,
 		return nil, errs.ErrClientGetProtoClient
 	}
 	resp, err := pdpb.NewPDClient(serviceClient.GetClientConn()).ScanRegions(cctx, req)
-	if !serviceClient.IsConnectedToLeader() && err != nil || resp.Header.GetError() != nil {
+	if serviceClient.NeedRetry(resp.GetHeader().GetError(), err) {
 		protoClient, cctx := c.getClientAndContext(scanCtx)
 		if protoClient == nil {
 			return nil, errs.ErrClientGetProtoClient
@@ -1245,8 +1245,8 @@ func (c *client) scatterRegionsWithGroup(ctx context.Context, regionID uint64, g
 	if err != nil {
 		return err
 	}
-	if resp.Header.GetError() != nil {
-		return errors.Errorf("scatter region %d failed: %s", regionID, resp.Header.GetError().String())
+	if resp.GetHeader().GetError() != nil {
+		return errors.Errorf("scatter region %d failed: %s", regionID, resp.GetHeader().GetError().String())
 	}
 	return nil
 }
@@ -1369,8 +1369,8 @@ func (c *client) scatterRegionsWithOptions(ctx context.Context, regionsID []uint
 	if err != nil {
 		return nil, err
 	}
-	if resp.Header.GetError() != nil {
-		return nil, errors.Errorf("scatter regions %v failed: %s", regionsID, resp.Header.GetError().String())
+	if resp.GetHeader().GetError() != nil {
+		return nil, errors.Errorf("scatter regions %v failed: %s", regionsID, resp.GetHeader().GetError().String())
 	}
 	return resp, nil
 }

--- a/tests/integrations/client/client_test.go
+++ b/tests/integrations/client/client_test.go
@@ -617,6 +617,15 @@ func (suite *followerForwardAndHandleTestSuite) TestGetTsoByFollowerForwarding1(
 	re.NoError(failpoint.Disable("github.com/tikv/pd/client/unreachableNetwork"))
 	time.Sleep(2 * time.Second)
 	checkTS(re, cli, lastTS)
+
+	re.NoError(failpoint.Enable("github.com/tikv/pd/client/responseNil", "return(true)"))
+	regions, err := cli.ScanRegions(ctx, []byte(""), []byte(""), 100)
+	re.NoError(err)
+	re.Len(regions, 0)
+	re.NoError(failpoint.Disable("github.com/tikv/pd/client/responseNil"))
+	regions, err = cli.ScanRegions(ctx, []byte(""), []byte(""), 100)
+	re.NoError(err)
+	re.Len(regions, 1)
 }
 
 // case 2: unreachable -> leader transfer -> normal

--- a/tests/integrations/client/client_test.go
+++ b/tests/integrations/client/client_test.go
@@ -621,7 +621,7 @@ func (suite *followerForwardAndHandleTestSuite) TestGetTsoByFollowerForwarding1(
 	re.NoError(failpoint.Enable("github.com/tikv/pd/client/responseNil", "return(true)"))
 	regions, err := cli.ScanRegions(ctx, []byte(""), []byte(""), 100)
 	re.NoError(err)
-	re.Len(regions, 0)
+	re.Empty(regions)
 	re.NoError(failpoint.Disable("github.com/tikv/pd/client/responseNil"))
 	regions, err = cli.ScanRegions(ctx, []byte(""), []byte(""), 100)
 	re.NoError(err)

